crates/pathfinder/src/p2p_network/sync_handlers/v1.rs

```diff
@@ -0,0 +1,579 @@
+                    block_number,
+                    class_hash,
+                    chunk,
+                    chunk_count,
+                    chunk_idx,
+                ));
+            }
+            // Now we reset the counter and start over with a clean slate
+            estimated_message_size = PER_MESSAGE_OVERHEAD;
+            // --> 1.
+        }
+    }
+
+    // 3. Send the remaining accumulated classes if there are any
+    if !classes_for_this_msg.is_empty() {
+        responses.push(block_bodies_response::take_from_class_definitions(
+            block_number,
+            &mut classes_for_this_msg,
+        ));
+    }
+    debug_assert!(classes_for_this_msg.is_empty());
+
+    Ok(())
+}
+
+mod block_bodies_response {
+    use p2p_proto_v1::block::BlockBodyMessage;
+
+    use super::*;
+
+    /// It is assumed that the chunk is not empty
+    pub fn from_class_definition_part(
+        block_number: BlockNumber,
+        class_hash: ClassHash,
+        part: &[u8],
+        parts_count: u32,
+        part_idx: u32,
+    ) -> BlockBodiesResponse {
+        use p2p_proto_v1::state::{Class, Classes};
+        BlockBodiesResponse {
+            block_number: block_number.get(),
+            body_message: BlockBodyMessage::Classes(Classes {
+                domain: 0, // FIXME
+                classes: vec![Class {
+                    compiled_hash: Hash(class_hash.0),
+                    definition: part.to_vec(),
+                    total_parts: Some(parts_count),
+                    part_num: Some(part_idx),
+                }],
+            }),
+        }
+    }
+
+    /// Pops all elements from `class_definitions` leaving the vector empty as if it was just `clear()`-ed,
+    /// so that later on it can be reused.
+    pub fn take_from_class_definitions(
+        block_number: BlockNumber,
+        class_definitions: &mut Vec<(ClassHash, Vec<u8>)>,
+    ) -> BlockBodiesResponse {
+        use p2p_proto_v1::state::{Class, Classes};
+        let classes = class_definitions
+            .drain(..)
+            .map(|(class_hash, definition)| Class {
+                compiled_hash: Hash(class_hash.0),
+                definition,
+                total_parts: None,
+                part_num: None,
+            })
+            .collect();
+        BlockBodiesResponse {
+            block_number: block_number.get(),
+            body_message: BlockBodyMessage::Classes(Classes {
+                domain: 0, // FIXME
+                classes,
+            }),
+        }
+    }
+}
+
+async fn spawn_blocking_get<Request, Response, Getter>(
+    request: Request,
+    storage: &Storage,
+    getter: Getter,
+) -> anyhow::Result<Response>
+where
+    Request: Send + 'static,
+    Response: Send + 'static,
+    Getter: FnOnce(Transaction<'_>, Request) -> anyhow::Result<Response> + Send + 'static,
+{
+    let storage = storage.clone();
+    let span = tracing::Span::current();
+
+    tokio::task::spawn_blocking(move || {
```

Comment:
actually rayon should be used here, but let's do that in a separate pr

Reply:
I don't think that's a good idea. Doing synchronous (file) I/O on the rayon thread pool is not a good idea: https://ryhl.io/blog/async-what-is-blocking/

Reply:
Ah yeah you're right, thanks